homeassistant/components/droplet/coordinator.py

```diff
@@ -0,0 +1,92 @@
+"""Droplet device data update coordinator object."""
+
+from __future__ import annotations
+
+import asyncio
+import logging
+import time
+
+from pydroplet.droplet import Droplet
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_CODE, CONF_HOST, CONF_NAME, CONF_PORT
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, RECONNECT_DELAY
+
+ML_L_CONVERSION = 1000
+VERSION_TIMEOUT = 5
+
+_LOGGER = logging.getLogger(__name__)
+
+
+type DropletConfigEntry = ConfigEntry[DropletDataCoordinator]
+
+
+class DropletDataCoordinator(DataUpdateCoordinator[None]):
+    """Droplet device object."""
+
+    config_entry: DropletConfigEntry
+
+    def __init__(self, hass: HomeAssistant, entry: DropletConfigEntry) -> None:
+        """Initialize the device."""
+        super().__init__(
+            hass, _LOGGER, config_entry=entry, name=f"{DOMAIN}-{entry.unique_id}"
+        )
+        self.droplet = Droplet(
+            host=entry.data[CONF_HOST],
+            port=entry.data[CONF_PORT],
+            token=entry.data[CONF_CODE],
+            session=async_get_clientsession(self.hass),
+            logger=_LOGGER,
+        )
+        self.dev_info = DeviceInfo()
+
+    async def _async_setup(self) -> None:
+        if not await self.setup():
+            raise ConfigEntryNotReady("Device is offline")
+
+        # Droplet should send its metadata within 5 seconds
+        end = time.time() + VERSION_TIMEOUT
+        assert self.config_entry.unique_id is not None
+        self.dev_info = DeviceInfo(
+            identifiers={(DOMAIN, self.config_entry.unique_id)},
+            name=self.config_entry.data[CONF_NAME],
+        )
```

Comment:
I would recommend to set the device info in the entity itself, as the coordinator is responsible for fetching data, and the entity platforms on how it will be exposed to HA.

Will you add more platforms in the future other than `sensor`? If so, you could do this in a base entity